controllers/storagecluster/reconcile.go

```diff
@@ -676,3 +690,146 @@ func validateOverprovisionControlSpec(sc *ocsv1.StorageCluster, reqLogger logr.L
+										},
+									},
+								},
+								{
+									Name: "ROOK_CEPH_SECRET",
+									ValueFrom: &corev1.EnvVarSource{
+										SecretKeyRef: &corev1.SecretKeySelector{
+											LocalObjectReference: corev1.LocalObjectReference{Name: "rook-ceph-mon"},
+											Key:                  "ceph-secret",
+										},
+									},
+								},
+							},
+							SecurityContext: &corev1.SecurityContext{
+								Privileged:   &privilegedContainer,
+								RunAsNonRoot: &runAsNonRoot,
+								RunAsUser:    &runAsUser,
+								RunAsGroup:   &runAsGroup,
+							},
+							VolumeMounts: []corev1.VolumeMount{
+								{Name: "ceph-config", MountPath: "/etc/ceph"},
+								{Name: "mon-endpoint-volume", MountPath: "/etc/rook"},
+							},
+						},
+					},
+					Tolerations: tolerations,
+					// if hostNetwork: false, the "rbd map" command hangs, see https://github.com/rook/rook/issues/2021
+					HostNetwork: true,
+					Volumes: []corev1.Volume{
+						{Name: "ceph-config", VolumeSource: corev1.VolumeSource{HostPath: &corev1.HostPathVolumeSource{Path: "/etc/ceph"}}},
+						{Name: "mon-endpoint-volume", VolumeSource: corev1.VolumeSource{
+							ConfigMap: &corev1.ConfigMapVolumeSource{LocalObjectReference: corev1.LocalObjectReference{Name: "rook-ceph-mon-endpoints"},
+								Items: []corev1.KeyToPath{
+									{Key: "data", Path: "mon-endpoints"},
+								},
+							},
+						},
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func (r *StorageClusterReconciler) ensureToolsDeployment(sc *ocsv1.StorageCluster) error {
+
+	var isFound bool
+	namespace := sc.Namespace
+
+	tolerations := []corev1.Toleration{{
+		Key:      defaults.NodeTolerationKey,
+		Operator: corev1.TolerationOpEqual,
+		Value:    "true",
+		Effect:   corev1.TaintEffectNoSchedule,
+	}}
+
+	tolerations = append(tolerations, sc.Spec.CephTolerations...)
+
+	toolsDeployment := newToolsDeployment(namespace, r.RookImage, tolerations)
+	foundToolsDeployment := &appsv1.Deployment{}
+	err := r.Client.Get(context.TODO(), types.NamespacedName{Name: rookCephToolDeploymentName, Namespace: namespace}, foundToolsDeployment)
+
+	if err == nil {
+		isFound = true
+	} else if errors.IsNotFound(err) {
+		isFound = false
+	} else {
+		return err
+	}
+
+	if sc.Spec.EnableCephTools {
+		// Create or Update if ceph tools is enabled.
+
+		//Adding Ownerreference to the ceph tools
+		err = controllerutil.SetOwnerReference(sc, toolsDeployment, r.Client.Scheme())
+		if err != nil {
+			return err
+		}
+
+		if !isFound {
+			return r.Client.Create(context.TODO(), toolsDeployment)
+		} else if !reflect.DeepEqual(foundToolsDeployment.Spec, toolsDeployment.Spec) {
+
+			updateDeployment := foundToolsDeployment.DeepCopy()
+			updateDeployment.Spec = *toolsDeployment.Spec.DeepCopy()
+
+			return r.Client.Update(context.TODO(), updateDeployment)
+		}
+	} else if isFound {
+		// delete if ceph tools exists and is disabled
+		return r.Client.Delete(context.TODO(), foundToolsDeployment)
+	}
```

Comment:
@malayparida2000 You still have an issue. 
As was discussed on this PR, we are going to keep support for the field on `OCSInitialization` as well.
Assuming that I am enabling with `OCSInitialization` the pod will get deployed, but then when the first reconcile of the storagecluster will happen the flag on the storagecluster will be false and the pod will get removed!

In your condition for deletion you need to also check that the value not set or set to false on the `OCSInitialization` CR

Reply:
@nb-ohad  For this I have to get the OCSInitialization CR inside the storagecluster reconciler & check for its value is set as false, But the vice versa also applies i.e I have to get the storage cluster CR inside ocsinitilization reconciler & before disabling the toolbox I have to check the spec for storagecluster CR. 
Or else it will create the same issue you mentioned above, i.e assuming that I am enabling with Storagecluster the pod will get deployed, but then when the first reconciliation of the ocsinitialization will happen the flag on the ocsinitialization will be false and the pod will get removed!
 So I was thinking Instead of this can I change the name of the toolbox that the storagecluster creates slightly? So that the cephtoolbox deployments both of them create are different from each other & both won't interfere with each other. Like the rookCephToolDeploymentName is set as "rook-ceph-tool" in the ocsinitialization, so can we set the name for the same in storagecluster something different, Like "rook-ceph-toolbox"?

Reply:
I don't agree, We should not have 2 toolbox pods at all.

I do understand that we can not remove changes from the ocsinitilization API, But we can remove the toolbox deployment code from the ocsinitilization controller and have this functionality in the storagecluster controller only.

In upgrades, we can check if it already exists and the flag is disabled in the storagecluster we will still update it. If the flag is enabled anyway we will create or update it.

Reply:
@malayparida2000  @jarrpa  @iamniting

This is more than just supporting the API, removing the functionality from the `ocsinitilization` controller will break existing autoamtions/documentation (not ours) - this includes the Managed Service! 
We need to have a timespan where both flags take effect to allow migration. 

I will start with saying that I am not agreeing with the main statement of this comment https://github.com/red-hat-storage/ocs-operator/pull/1602#discussion_r861282028 and I will explain:

Assuming that we are ok with watching and reading the `ocsinitilization` CR in the `stoagecluster` controller for the next release then we can do the following:
1. the `storagecluster` controller will be the one to deploy the toolbox!
2. the condition for deployment will be an OR operation between the enablement flags on both CRs.
3. the `storagecluster` will watch the `ocsinitilization` CR (which you will need for upgrades anyway) which will trigger a reconcile if the flag changes on `ocsinitilization` CR

The main thing here is that we do not need to sync both CRs, we enable if either one is on and we disable only if both are off!

Running with this code in place will give us the opportunity to deprecate the feature from the `ocsinitilization` controller during a full release, and on the release that comes after it, we can just ignore the flag on `ocsinitilization` CR completely assuming that we gave enough time and warning to our customers to update their automation/documentation

Reply:
I believe the results of this discussion have been met. @nb-ohad PTAL

Reply:
Seems Ok